pytorch_lightning/loops/epoch/training_epoch_loop.py

```diff
@@ -146,18 +145,16 @@ def advance(self, *args: Any, **kwargs: Any) -> None:
         if self._num_training_batches_reached():
             self.update_lr_schedulers("epoch", update_plateau_schedulers=False)
 
-        batch_end_outputs = [opt_idx_out for opt_idx_out in batch_output.training_step_output if len(opt_idx_out)]
-        processed_batch_end_outputs = self._prepare_outputs(batch_end_outputs, batch_mode=True)
-
-        # hook
-        self.trainer.call_hook("on_train_batch_end", processed_batch_end_outputs, batch, self.batch_idx, 0)
+        if is_overridden("on_train_batch_end", self.trainer.lightning_module):
+            batch_end_outputs = self._prepare_outputs_training_batch_end(batch_output.outputs)
+            self.trainer.call_hook("on_train_batch_end", batch_end_outputs, batch, self.batch_idx, 0)
```

Comment:
I think we need to call this without the overridden check, because the checkpoint callback for example relies on this. do you remember why we did this?

Reply:
Ahh good eye.

We did this to avoid the computation when it's not overridden

But forgot that `call_hook` calls everything.

I can try and address this with #9029 - we can undo the check here.

Reply:
It's unfortunate that `ModelCheckpoint` uses `on_train_batch_end` since it doesn't use the outputs at all. This would be one pro for #8731